src/garage/torch/algos/bc.py

```diff
@@ -0,0 +1,190 @@
+"""Implementation of Behavioral Cloning in PyTorch."""
+import itertools
+
+from dowel import tabular
+import numpy as np
+import torch
+
+from garage import _Default, log_performance, make_optimizer, TimeStepBatch, \
+    TrajectoryBatch
+from garage.np import obtain_evaluation_samples
+from garage.np.algos.rl_algorithm import RLAlgorithm
+from garage.sampler import RaySampler
+from garage.torch import np_to_torch
+from garage.torch.policies import Policy
+
+
+class BC(RLAlgorithm):
+    """Behavioral Cloning.
+
+    Based on Model-Free Imitation Learning with Policy Optimization:
+        https://arxiv.org/abs/1605.08478
+
+    Args:
+        env_spec (garage.envs.EnvSpec): Specification of environment.
+        learner (garage.torch.Policy): Policy to train.
+        batch_size (int): Size of optimization batch.
+        source (garage.Policy or Generator[garage.TimeStepBatch]): Expert to
```

Comment:
out-of-scope for this PR: what do you think about defining a Dataset API, which would unify things like this (in addition to accessing the replay buffer, or on-policy sampling).

basically `TimeStepDataset = Generator[garage.TimeStepBatch]` and `TrajectoryDataset = Generator[garage.TrajectoryBatch]`

alternatively, do this but use the torch and tf first-party APIs

Reply:
Yeah, I'm not sure if the first party APIs are best to use or not at this point. It seems unlikely we'd go wrong using a more restrictive API for the mean time.